Guess offsets

iaguis opened the pull request:
Instead of requiring a different ebpf object file per kernel version /
distro, we try to guess the offsets of the fields we access in the eBPF
program.

We do so by creating connections to a server we start, trying to read
through struct sock, and comparing the obtained values with the expected
ones.

Requires https://github.com/kinvolk/tcptracer-bpf/pull/15

iaguis committed:
Guess offsets

Instead of requiring a different ebpf object file per kernel version /
distro, we try to guess the offsets of the fields we access in the eBPF
program.

We do so by creating connections to a server we start, trying to read
through struct sock, and comparing the obtained values with the expected
ones.

alban commented:
@schu @alepuccetti: The changes in bpf/bpf.go would go to iovisor/gobpf but the changes in main.go would go to Scope.

alban commented on bpf/bpf.go:
```diff
@@ -360,6 +437,54 @@ func (b *BPFKProbePerf) readVersion() (int, error) {
+
+	releaseStr := strings.Trim(utsnameStr(buf.Release[:]), "\x00")
+
+	kernelVersionStr := strings.Split(releaseStr, "-")[0]
```
@iaguis This does not work in the rkt kernel (rkt with the kvm flavor) because releaseStr is "4.8.6rkt-v1":
```
# uname -r
4.8.6rkt-v1
```

Symptoms:
```
strconv.ParseInt: parsing "6rkt": invalid syntax
```

alban replied:
tested with https://github.com/coreos/rkt/pull/3487

schu replied:
Doesn't break on my Debian box but the result doesn't reflect the current kernel version it seems: `uname -r` is `4.8.0-2-amd64` (kernel version `264192`) whereas the actual version seems to be `4.8.11` (`264203`) and the prog cannot be loaded..

alban replied:
The fix in rkt about localversion got merged https://github.com/coreos/rkt/pull/3489

alban replied:
@schu are you testing the correct branch of tcptracer-bpf? https://github.com/kinvolk/tcptracer-bpf/pull/15

schu replied:
@alban I'm testing with [dummy.c](https://github.com/iovisor/gobpf/blob/ad48fe817a36356d8c1d206c9939d992a1243e2c/tests/dummy.c) from https://github.com/iovisor/gobpf/pull/6

alban commented:
Under the rkt (flavor kvm) kernel, this guesses the offsets correctly most of the time, but sometimes it freezes. dmesg says:
```
nf_conntrack: nf_conntrack: table full, dropping packet
```
We should add a timeout, so it does not wait for too long for the connection to succeed.

Also, this code is not supposed to do many connections simultaneously, so this should not be a problem for conntrack. We might not close the connection correctly before testing the next offset.

schu commented:
> The changes in bpf/bpf.go would go to iovisor/gobpf

As gobpf now includes our elf package, the bpf part can be a separate pull request against iovisor/gobpf.

iaguis commented:
> Also, this code is not supposed to do many connections simultaneously, so this should not be a problem for conntrack. We might not close the connection correctly before testing the next offset.

Maybe we fill it up because the closed connections stay in `TIME_WAIT` or such?

iaguis commented:
> Maybe we fill it up because the closed connections stay in TIME_WAIT or such?

I just tried with rkt-kvm and the connections stay in `TIME_WAIT` even after the program exits:

```
root@rkt-46288066-3418-44b3-8854-a1e9980466d5:/# conntrack -C
17

*** run ./gobpf-elf-loader ./ebpf.o

root@rkt-46288066-3418-44b3-8854-a1e9980466d5:/# conntrack -C
858

*** stop ./gobpf-elf-loader

root@rkt-46288066-3418-44b3-8854-a1e9980466d5:/# conntrack -L | grep TIME_WAIT | wc -l
conntrack v1.4.2 (conntrack-tools): 854 flow entries have been shown.
845
```

iaguis committed:
Set SO_LINGER to 0

We make a lot of connections when we get the offsets and they can
potentially fill the conntrack table because the state the connections
remain in when they get closed is TIME_WAIT and it stays like that for a
while (around 2 minutes).

We set SO_LINGER to 0 so the connection state after closing is CLOSE
instead. In this way, they will disappear from the conntrack table after
around 10 seconds.

alban commented on main.go:
```diff
@@ -298,6 +298,13 @@ func guessOffsets(b *bpf.BPFKProbePerf) error {
 			}
+
+			if tcpConn, ok := conn.(*net.TCPConn); ok {
+				tcpConn.SetLinger(0)
```
A comment would be nice to explain why

iaguis replied:
done

iaguis committed:
Switch to iovisor/gobpf

iaguis commented:
kinvolk/tcptracer-bpf#15 is merged. Let's merge this.
